src/lightning_app/components/training.py

```diff
@@ -0,0 +1,167 @@
+        node_rank: int = 1,
+        num_nodes: int = 1,
+        sanity_serving: bool = False,
+        cloud_compute: Optional[CloudCompute] = None,
+        **kwargs,
+    ):
+        super().__init__(
+            script_path, script_args, raise_exception=True, parallel=True, cloud_compute=cloud_compute, **kwargs
+        )
+        self.node_rank = node_rank
+        self.num_nodes = num_nodes
+        self.best_model_path = None
+        self.best_model_score = None
+        self.sanity_serving = sanity_serving
+        self.has_finished = False
+
+    def configure_tracer(self):
+        from pytorch_lightning import Trainer
+
+        tracer = super().configure_tracer()
+        tracer.add_traced(Trainer, "__init__", pre_fn=self._trainer_init_pre_middleware)
+        return tracer
+
+    def run(self, internal_urls: Optional[List[Tuple[str, str]]] = None):
+        if not internal_urls:
+            _logger.info(f"The node {self.node_rank} started !")
+            return
+
+        master_address = str(internal_urls[0][0])
+        master_port = str(internal_urls[0][1])
+
+        os.environ.update(
+            {
+                "MASTER_ADDR": master_address,
+                "MASTER_PORT": master_port,
+                "NODE_RANK": str(self.node_rank),
+                "WORLD_SIZE": str(self.num_nodes * self.cloud_compute.devices),
+                "PL_TRAINER_NUM_NODES": str(self.num_nodes),
+                "PL_TRAINER_STRATEGY": "ddp",
+                "PL_TRAINER_DEVICES": str(self.cloud_compute.devices),
+                "PL_TRAINER_ACCELERATOR": "auto",
+            }
+        )
+        return super().run()
+
+    def on_after_run(self, script_globals):
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.utilities.cli import LightningCLI
+
+        cli = [v for v in script_globals.values() if isinstance(v, LightningCLI)]
+        if cli:
+            trainer = cli[0].trainer
+        else:
+            trainer = [v for v in script_globals.values() if isinstance(v, Trainer)][0]
+
+        if trainer.checkpoint_callback.best_model_score:
+            self.best_model_path = Path(trainer.checkpoint_callback.best_model_path)
+            self.best_model_score = float(trainer.checkpoint_callback.best_model_score)
+        self.has_finished = True
+        # TODO: Why does it hang there.
+        raise SystemExit(0)
+
+    def _trainer_init_pre_middleware(self, trainer, *args, **kwargs):
+        if self.node_rank != 0:
+            return {}, args, kwargs
+
+        from pytorch_lightning.serve import ServableModuleValidator
+
+        callbacks = kwargs.get("callbacks", [])
+        if self.sanity_serving:
+            callbacks = callbacks + [ServableModuleValidator()]
+        kwargs["callbacks"] = callbacks
+        return {}, args, kwargs
+
+    @property
+    def is_running_in_cloud(self) -> bool:
+        return "LIGHTNING_APP_STATE_URL" in os.environ
```

Comment:
I'm not sure this is a correct assumption. Could it be 
```python 
bool(os.environ.get("LIGHTNING_CLOUD_APP_ID", False))
```

Reply:
I think both are fine. I don't have a preference.

Reply:
is `LIGHTNING_APP_STATE_URL` set locally ?